### PR TITLE
feat: implement grc1155

### PR DESCRIPTION
**Pull request by Jammyaa**
# Description
- Implement GRC1155 base methods
- Implement foo1155 realm based on GRC1155
- Add tests for p/demo/grc/grc1155 and r/demo/foo1155

# How has this been tested?
done all cases for p/demo/grc/grc1155 and r/demo/foo1155

### CODE REVIEW

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:229]: since we're not reusing event, `emit(&TransferSingleEvent{caller, from, to, tid, amount})`? (here as well as the occurrences below)

File: examples/gno.land/p/demo/grc/grc1155/util.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/util.gno:14]: maybe change this to return a bool instead of an error? this way, the callers can do the following

```go
if !isValidAddress(from) || !isValidAddress(to) {
    return ErrInvalidAddress
}
```

in case we should distinguish the errors we can always change this as this is unexported.

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:234]: what if, to avoid duplicating logic, this entire function was just:

```go
// called from SafeTransferFrom
s.safeBatchTransferFrom(from, to, []TokenID{tid}, []uint64{amount})
```

and the event emission was pulled out of `safeTransferFrom` and batch version, and were put instead in their exported counterparts? (so each function actually emits the correct event)

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:310]: again, I think for reducing code and logic this function could be removed and the `Mint` method could instead just call `mintBatch` with the arrays, making sure that the events are emitted accordingly.

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:372]: same for `Burn` and `burnBatch`

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:410]: Should we not emit `UpdateURIEvent`?

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno:11]: const?

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno:390]: please, in general, add some more tests regarding error conditions. for instance:

* should not be able to get the balance of an invalid address
* should not be able to transfer tokens if my balance is less than what i request
* giving bad inputs to batch operations fails (len(addr) != len(amount)), and giving them nothing is a no-op.

File: examples/gno.land/p/demo/grc/grc1155/util.gno
Comment (thehowl) [inline examples/gno.land/p/demo/grc/grc1155/util.gno:7]: since the Address type is just a string, `const zeroAddress std.Address = ""`?

**Review by thehowl — CHANGES_REQUESTED** [review]
Thank you for your contribution! I've found a few things to change and some places where the code can be greatly simplified and reduced in size, thank you in advance for addressing these issues :)

**Comment by thehowl** [comment]
You'll also need to add the package and the realm to `cmd/gnoland/main.go` so that they're loaded on the staging and any testnets. :)

**Comment by Jammyaa** [comment]
Ok, I will resolve all issues. Thank you for your quality comment!

**Review by thehowl — COMMENTED** [review]
rest looks good :)

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (Jammyaa) [reply examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:19]: Ok, I will consider this suggestion. Thank you.

Comment (Jammyaa) [reply examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:19]: Sorry, but if you modify this return type from *basicGRC1155Token by IGRC1155, you will see the panic in your terminal. 

And if you see p/demo/grc721 of @harry-hov, he uses *basicNFT (not IGRC721) in the return type of NewBasicNFT() method, similar to me. 

One more thing, I think that should not expose basicGRC1155Token and basicNFT struct to end user directly, so I use lowercase to make it only be accessed within the grc1155 package.

So what's the problem with using *basicGRC1155Token or *basicNFT? Thank you very much @thehowl

Comment (thehowl) [reply examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:19]: Hmm... I've reviewed and tried putting the interface. The tests you've made kind of rely on the unexported methods, so it makes sense to keep things like this at the moment. Ideally, we'd still get into a state where in a testing environment we can manipulate the caller address, and refactor the code that currently returns an unexported struct when they should return an interface after that.

The reasoning why I personally avoid returning unexported structs is that in the context of normal Go, with all the tooling etc., from a godoc page you can only see the methods of the exported types. Of course, tooling still knows how to work with unexported types, so for instance `gopls` and other autocompletion tools will still give you documentation for the methods of an unexported type, but in general while it makes it easier to go look up the implementation code of the type, it makes it harder to go look up the documentation from a godoc page.

I'll create a separate issue for this, though.

**Review by thehowl — APPROVED** [review]
looks good to me, thank you for your time and patience.

Pinging @moul for a last review :)

**Comment by Jammyaa** [comment]
Hi sir, I have addressed all issues that Morgan comments, and got his approval, so could you help me review the last time and merge this code? Thank you very much @moul

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:341]: Are these leftover?

**Review by zivkovicmilos — APPROVED** [review]
Thank you for the contribution 🙏 

Looks good 💯

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno
Comment (zivkovicmilos) [reply examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:341]: Got it, no problem, just wanted to double check

**Comment by Jammyaa** [comment]
Excuse me, I don't know if there is still a problem with this contribution of mine, it has not been merged after being approved for a long time. 
I am continuing my work by implementing soul bound token for Gnoland, can someone merge and close the PR so that I can continue to contribute from my fork branch. Thank you very much. @moul @thehowl

**Review by moul — APPROVED** [review]

**Pull request merged**
